### PR TITLE
add whitebox test

### PR DESCRIPTION
**Pull request by liufengyun**
During meeting with @olafurpg , we checked that whitebox macros work in Gestalt.

### CODE REVIEW

File: src/main/scala/gestalt/api.scala [hunk 1]
```diff
@@ -411,6 +411,7 @@ object api extends Toolbox {
   }
 
   implicit def tpd2untpd(tree: tpd.Tree): Splice = TypedSplice(tree)
+  // implicit def tpds2untpds(ts: List[tpd.Tree]): List[Tree] = ts.map(_.wrap)
```
Comment (olafurpg) [inline src/main/scala/gestalt/api.scala]: Remove?

File: macros/src/test/scala/gestalt/macros/DefMacroTest.scala [hunk 2]
```diff
@@ -211,4 +211,9 @@ class DefMacroTest extends TestSuite {
     assert(trees.some3Explicit() == Some(3))
     assert(trees.some3ExplicitTyped() == Some(3))
   }
+
+  test("white box") {
+    val x = Whitebox.gimmelist(2)
```
Comment (xeno-by) [inline macros/src/test/scala/gestalt/macros/DefMacroTest.scala]: What is the type of `x` here? Can one annotate it with `List[Int]`?

**Comment by xeno-by** [comment]
Whitebox macros have other aspects that also include: 1) dynamically computing availability of implicits, 2) assisting type inference, 3) customization of pattern matching (http://docs.scala-lang.org/overviews/macros/blackbox-whitebox.html).

File: macros/src/test/scala/gestalt/macros/DefMacroTest.scala (hunk 2 above)
Comment (olafurpg) [inline macros/src/test/scala/gestalt/macros/DefMacroTest.scala]: Yes, the type of `x` is List[Int] here.

Comment (liufengyun) [inline macros/src/test/scala/gestalt/macros/DefMacroTest.scala]: Just tried locally, explicit annotation doesn't work, which is surprising to programmers. While it's possible to remove the surprise, it's only worth the effort if whitebox should be supported.

Comment (xeno-by) [inline macros/src/test/scala/gestalt/macros/DefMacroTest.scala]: Interesting. Why?

**Comment by olafurpg** [comment]
It would be interesting @liufengyun to explore implementing in Dotty a macro like `materializeIso` from here https://github.com/scala/scala/pull/2821/files#diff-bb0779431873314ea7f49688bd157d90R11. See "Fundep materialization" http://docs.scala-lang.org/overviews/macros/implicits.html#fundep_materialization for context

File: macros/src/test/scala/gestalt/macros/DefMacroTest.scala (hunk 2 above)
Comment (liufengyun) [inline macros/src/test/scala/gestalt/macros/DefMacroTest.scala]: With/without explicit type annotation, the type checking go through different steps. If without, `x` gets a type from the rhs, which is expanded and type checked as `List[Int]`.

With expected type `List[Int]`, there is an assert before macro expansion, thus compilation failed. We could change the implementation to try to expand macros before fail with expected type.

Anyway, I had the impression that whitebox doesn't have much to do with macro system design. It's more a language design decision.

**Comment by liufengyun** [comment]
Thanks for the pointer @olafurpg . I'm happy to play with that. But indeed whiteboxity is more a language design decision, it doesn't influence the design of the macro system (user and compiler APIs). I think it's fine to postpone the decision for now.

**Comment by xeno-by** [comment]
Whether or not our new macro system will support whiteboxity is a contention point for many macro users. 

We can postpone the actual decision a little bit until our first milestone is ready for release, but I think it is urgent to figure out whether Dotty can theoretically support whiteboxity. If it can't (e.g. if type inference is organized in a way that precludes implicit macros from guiding it), then it's a significant problem that we need to realize early.

**Comment by liufengyun** [comment]
Current Dotty compiler enforces instantiation of type vars at the boundary of DefTrees:

```Scala
  def adapt(tree: Tree, pt: Type, original: untpd.Tree = untpd.EmptyTree)(implicit ctx: Context): Tree = /*>|>*/ track("adapt") /*<|<*/ {
    /*>|>*/ ctx.traceIndented(i"adapting $tree of type ${tree.tpe} to $pt", typr, show = true) /*<|<*/ {
      if (tree.isDef) interpolateUndetVars(tree, tree.symbol)
      else if (!tree.tpe.widen.isInstanceOf[LambdaType]) interpolateUndetVars(tree, NoSymbol)
      tree.overwriteType(tree.tpe.simplified)
      adaptInterpolated(tree, pt, original)
    }
  }
```

Also, implicit resolution happens immediately when it's required. This means we'll have an error at the line `val tp` (`L` is inferred to be `Nothing`).

```Scala
trait Iso[T, U] {
  def to(t: T) : U
  def from(u: U) : T
}
case class Foo(i: Int, s: String, b: Boolean)
def conv[C, L](c: C)(implicit iso: Iso[C, L]): L = iso.to(c)
val tp  = conv(Foo(23, "foo", true))      //   <====   error
tp: (Int, String, Boolean)
tp == (23, "foo", true)
```

Generally, I don't think Dotty is friendly for delaying implicit resolution by taking into consideration of some type constraints that appear later in the code.

**Pull request merged**